consul/server_manager/server_manager.go

```diff
@@ -0,0 +1,340 @@
+	"log"
+	"math/rand"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/hashicorp/consul/consul/server_details"
+	"github.com/hashicorp/consul/lib"
+)
+
+type consulServerEventTypes int
+
+const (
+	// clientRPCJitterFraction determines the amount of jitter added to
+	// clientRPCMinReuseDuration before a connection is expired and a new
+	// connection is established in order to rebalance load across consul
+	// servers.  The cluster-wide number of connections per second from
+	// rebalancing is applied after this jitter to ensure the CPU impact
+	// is always finite.  See newRebalanceConnsPerSecPerServer's comment
+	// for additional commentary.
+	//
+	// For example, in a 10K consul cluster with 5x servers, this default
+	// averages out to ~13 new connections from rebalancing per server
+	// per second (each connection is reused for 120s to 180s).
+	clientRPCJitterFraction = 2
+
+	// clientRPCMinReuseDuration controls the minimum amount of time RPC
+	// queries are sent over an established connection to a single server
+	clientRPCMinReuseDuration = 120 * time.Second
+
+	// initialRebalanceTimeoutHours is the initial value for the
+	// rebalanceTimer.  This value is discarded immediately after the
+	// client becomes aware of the first server.
+	initialRebalanceTimeoutHours = 24
+
+	// Limit the number of new connections a server receives per second
+	// for connection rebalancing.  This limit caps the load caused by
+	// continual rebalancing efforts when a cluster is in equilibrium.  A
+	// lower value comes at the cost of increased recovery time after a
+	// partition.  This parameter begins to take effect when there are
+	// more than ~48K clients querying 5x servers or at lower server
+	// values when there is a partition.
+	//
+	// For example, in a 100K consul cluster with 5x servers, it will
+	// take ~5min for all servers to rebalance their connections.  If
+	// 99,995 agents are in the minority talking to only one server, it
+	// will take ~26min for all servers to rebalance.  A 10K cluster in
+	// the same scenario will take ~2.6min to rebalance.
+	newRebalanceConnsPerSecPerServer = 64
+)
+
+type ConsulClusterInfo interface {
+	NumNodes() int
+}
+
+// serverCfg is the thread-safe configuration structure that is used to
+// maintain the list of consul servers in Client.
+//
+// NOTE(sean@): We are explicitly relying on the fact that serverConfig will
+// be copied onto the stack.  Please keep this structure light.
+type serverConfig struct {
+	// servers tracks the locally known servers.  List membership is
+	// maintained by Serf.
+	servers []*server_details.ServerDetails
+}
+
+type ServerManager struct {
+	// serverConfig provides the necessary load/store semantics for the
+	// server list.
+	serverConfigValue atomic.Value
+	serverConfigLock  sync.Mutex
+
+	// shutdownCh is a copy of the channel in consul.Client
+	shutdownCh chan struct{}
+
+	logger *log.Logger
+
+	// serf is used to estimate the approximate number of nodes in a
```

Comment:
Might want to mention that `clusterInfo` wraps Serf. This is probably just stale.